File: hackertracker/src/main/java/com/shortstack/hackertracker/View/FilterView.java
package com.shortstack.hackertracker.View;

import android.content.Context;
import android.content.res.ColorStateList;
import android.support.v4.widget.CompoundButtonCompat;
import android.support.v7.widget.AppCompatCheckBox;
import android.util.AttributeSet;
import android.widget.CheckBox;
import android.widget.LinearLayout;

import com.shortstack.hackertracker.Application.App;
import com.shortstack.hackertracker.Database.DatabaseController;
import com.shortstack.hackertracker.Model.Filter;
import com.shortstack.hackertracker.Model.Types;
import com.shortstack.hackertracker.R;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class FilterView extends LinearLayout {

    AppCompatCheckBox[] checkboxes;

    @BindView(R.id.filter_left)
    LinearLayout left;

    @BindView(R.id.filter_right)
    LinearLayout right;

    public FilterView(Context context) {
        super(context);
        init();
    }

    public FilterView(Context context, Filter filter) {
        super(context);
        init();
        setFilter(filter);
    }

    private void setFilter(Filter filter) {
        String[] typesArray = filter.getTypesArray();

        if( typesArray.length == 0 ) {
            for (AppCompatCheckBox type : checkboxes) {
                type.setChecked(true);
            }
        }

        for (String aTypesArray : typesArray) {
            for (int i1 = 0; i1 < checkboxes.length; i1++) {
                if (aTypesArray.equals(checkboxes[i1].getText().toString())) {
                    checkboxes[i1].setChecked(true);
                }
            }
        }
    }

    public FilterView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public FilterView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        inflate(getContext(), R.layout.alert_filter, this);
        ButterKnife.bind(this);

        DatabaseController controller = App.application.getDatabaseController();
        List<Types.Type> types = controller.getTypes();

        checkboxes = new AppCompatCheckBox[types.size()];

        int[] stringArray = getContext().getResources().getIntArray(R.array.colors);

        int states[][] = {{android.R.attr.state_checked}, {}};

        for (int i = 0; i < types.size(); i++) {
            Types.Type type = types.get(i);

            AppCompatCheckBox box = new AppCompatCheckBox(getContext());
            box.setText(type.getType());
            CompoundButtonCompat.setButtonTintList(box, new ColorStateList(states, new int[]{stringArray[i], stringArray[i]}));

            if (i < types.size() / 2)
                left.addView(box);
            else
                right.addView(box);

            checkboxes[i]= box;
        }


    }

    public Filter save() {
        ArrayList<String> selected = new ArrayList<>();

        for (int i = 0; i < checkboxes.length; i++) {
            CheckBox type = checkboxes[i];
            if (type.isChecked()) {
                selected.add(checkboxes[i].getText().toString());
            }
        }

        String[] strings = selected.toArray(new String[selected.size()]);

        Filter filter = new Filter(strings);

        App.Companion.getStorage().saveFilter(filter);

        return filter;
    }
}

File: hackertracker/src/main/java/com/shortstack/hackertracker/View/FilterView.kt
package com.shortstack.hackertracker.view

import android.arch.persistence.room.Room
import android.content.Context
import android.content.res.ColorStateList
import android.support.v4.widget.CompoundButtonCompat
import android.support.v7.widget.AppCompatCheckBox
import android.util.AttributeSet
import android.view.View
import android.widget.LinearLayout
import com.shortstack.hackertracker.App
import com.shortstack.hackertracker.R
import com.shortstack.hackertracker.database.MyRoomDatabase
import com.shortstack.hackertracker.models.Filter
import com.shortstack.hackertracker.models.Type
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.alert_filter.view.*
import java.util.*

class FilterView : LinearLayout {

    lateinit var checkboxes: Array<AppCompatCheckBox>

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, filter: Filter) : super(context) {
        init()
//        setFilter(filter)
    }

    private fun setFilter(filter: Filter) {
        val typesArray = filter.typesArray

        if (typesArray.size == 0) {
            for (type in checkboxes) {
                type.isChecked = true
            }
        }

        for (aTypesArray in typesArray) {
            for (i1 in checkboxes.indices) {
                if (aTypesArray == checkboxes[i1].text.toString()) {
                    checkboxes[i1].isChecked = true
                }
            }
        }
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()


    }

    private fun init() {
        View.inflate(context, R.layout.alert_filter, this)

        App.application.database.db.typeDao().getTypes()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ types ->
                    setCheckboxes(types)
                }, {

                })
    }

    private fun setCheckboxes(types: List<Type>) {
        val stringArray = context.resources.getIntArray(R.array.colors)

        val states = arrayOf(intArrayOf(android.R.attr.state_checked), intArrayOf())

        checkboxes = Array(types.size, {
            val type = types[it]

            val box = AppCompatCheckBox(context)
            box.text = type.type
            CompoundButtonCompat.setButtonTintList(box, ColorStateList(states, intArrayOf(stringArray[it], stringArray[it])))

            box.isChecked = type.isSelected

            if (it <= types.size / 2)
                filter_left.addView(box)
            else
                filter_right.addView(box)

            return@Array box
        })
    }

    fun save(): Filter {
        val selected = ArrayList<String>()

        for (i in checkboxes.indices) {
            val type = checkboxes[i]
            if (type.isChecked) {
                selected.add(checkboxes[i].text.toString())
            }
        }

        val strings = selected.toTypedArray()

        val filter = Filter(strings)

        App.application.storage.saveFilter(filter)

        return filter
    }
}
